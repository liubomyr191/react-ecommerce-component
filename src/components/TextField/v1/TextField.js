import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { applyTheme } from "./../../../helpers";

const StyledInput = styled.input`
  background-color: ${(props) => {
    let inputType;
    if (props.dark) {
      inputType = "dark"
    } else {
      inputType = "default"
    }
    return  applyTheme(`inputBackgroundColor_${inputType}`)(props);
}};
  &:focus {
    background-color: ${applyTheme("inputBackgroundColor_default_focus")};
  }
`;

const StyledTextarea = styled.textarea`
  background-color: #ff00ff;
`;

class TextField extends Component {
  static isFormInput = true;

  static propTypes = {
    allowLineBreaks: PropTypes.bool,
    className: PropTypes.string,
    convertEmptyStringToNull: PropTypes.bool,
    isReadOnly: PropTypes.oneOfType([PropTypes.bool, PropTypes.func]),
    maxLength: PropTypes.number,
    name: PropTypes.string,
    onChange: PropTypes.func,
    onChanging: PropTypes.func,
    onSubmit: PropTypes.func,
    placeholder: PropTypes.string,
    trimValue: PropTypes.bool,
    type: PropTypes.oneOf([
      "email",
      "password",
      "text",
      "url",
    ]),
    value: PropTypes.string
  };

  static defaultProps = {
    allowLineBreaks: false,
    convertEmptyStringToNull: true,
    isReadOnly: false,
    onChange() {},
    onChanging() {},
    onSubmit() {},
    trimValue: true,
    type: "text"
  };

  constructor(props) {
    super(props);

    this.state = {
      value: props.value || ""
    };
  }

  componentWillMount() {
    const { value } = this.state;
    this.handleChanging(value);
    this.handleChanged(value);
  }

  componentWillReceiveProps(nextProps) {
    const { value } = this.props;
    const { value: nextValue } = nextProps;

    // Whenever a changed value prop comes in, we reset state to that, thus becoming clean.
    if (value !== nextValue) {
      this.setValue(nextValue);
    }
  }

  onKeyPress = (event) => {
    const { onSubmit } = this.props;
    if (event.which === 13) onSubmit();
  };

  onBlur = (event) => {
    this.setValue(event.target.value);
  };

  onChange = (event) => {
    let { value } = event.target;
    value = value || "";
    this.setState({ value });
    this.handleChanging(value);
  };

  getValue() {
    return this.cleanValue(this.state.value);
  }

  setValue(value = "") {
    this.setState({ value });
    this.handleChanging(value);
    this.handleChanged(value);
  }

  cleanValue(value) {
    const { convertEmptyStringToNull, trimValue } = this.props;
    let outputValue = trimValue ? value.trim() : value;
    if (convertEmptyStringToNull && outputValue === "") outputValue = null;
    return outputValue;
  }

  resetValue() {
    this.setValue(this.props.value);
  }

  handleChanged(value) {
    const { onChange } = this.props;
    const outputValue = this.cleanValue(value);
    if (outputValue !== this.lastChangedValue) {
      this.lastChangedValue = outputValue;
      onChange(outputValue);
    }
  }

  handleChanging(value) {
    const { onChanging } = this.props;
    const outputValue = this.cleanValue(value);
    if (outputValue !== this.lastChangingValue) {
      this.lastChangingValue = outputValue;
      onChanging(outputValue);
    }
  }

  // Input is dirty if value prop doesn"t match value state. Whenever a changed
  // value prop comes in, we reset state to that, thus becoming clean.
  isDirty() {
    return this.state.value !== this.props.value;
  }

  render() {
    const { allowLineBreaks, className, isReadOnly, maxLength, name, placeholder, type } = this.props;
    const { value } = this.state;

    if (allowLineBreaks) {
      // Same as "input" but without `onKeyPress` and `type` props.
      // We don"t support rows; use style to set height instead
      return (
          <StyledTextarea
            className={className}
            readOnly={isReadOnly}
            maxLength={maxLength}
            name={name}
            onBlur={this.onBlur}
            onChange={this.onChange}
            placeholder={placeholder}
            value={value}
          />
      );
    }

    return (
        <StyledInput
          className={className}
          readOnly={isReadOnly}
          maxLength={maxLength}
          name={name}
          onKeyPress={this.onKeyPress}
          onBlur={this.onBlur}
          onChange={this.onChange}
          placeholder={placeholder}
          type={type}
          value={value}
        />
    );
  }
}

export default TextField;
