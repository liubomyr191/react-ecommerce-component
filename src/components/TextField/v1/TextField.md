### TextField Overview

##### Types
Single line, text area.

```jsx
<div style={{ display: "flex" }}>
  <div style={{ border: "1px solid #cccccc", flex: "0 1 50%", padding: "1rem" }}>
    <TextField name="single-line" value="VALUE" placeholder="Hint" errors={["error"]} />
  </div>
</div>
``` 

#### Single line text field

##### Types
Text field on white and text field on grey or dark backgrounds

```jsx
<div style={{ display: "flex" }}>
  <div style={{ border: "1px solid #cccccc", flex: "1 1 auto", padding: "1rem" }}>
    <TextField name="single-line" placeholder="Hint" icon={<i className="fa fa-wifi"></i>} />
  </div>
  <div style={{ backgroundColor: "#f5f5f5", flex: "1 1 auto", margin: "0 1rem", padding: "1rem" }}>
    <TextField name="single-line" placeholder="Hint" dark />
  </div>
</div>
```

#### Single line text field states

##### States
Text fields have 8 basic states idle unfilled, focused, active submit, idle and filled in, filled and focused, validated, error, and disabled.

```jsx
<div style={{ display: "flex" }}>
  <div style={{ border: "1px solid #cccccc", flex: "1 1 auto", padding: "1rem" }}>
    <TextField name="single-line" placeholder="Hint" />
  </div>
  <div style={{ backgroundColor: "#f5f5f5", flex: "1 1 auto", margin: "0 1rem", padding: "1rem" }}>
    <TextField name="single-line" placeholder="Hint" dark />
  </div>
</div>
```

#### Multi line text field

##### Types
Text field on white and text field on grey or dark backgrounds

```jsx
<div style={{ display: "flex" }}>
  <div style={{ border: "1px solid #cccccc", flex: "1 1 auto", padding: "1rem" }}>
    <TextField name="multi-line" placeholder="Hint" allowLineBreaks />
  </div>
  <div style={{ backgroundColor: "#f5f5f5", flex: "1 1 auto", margin: "0 1rem", padding: "1rem" }}>
    <TextField name="multi-line" placeholder="Hint" dark allowLineBreaks />
  </div>
</div>
```
