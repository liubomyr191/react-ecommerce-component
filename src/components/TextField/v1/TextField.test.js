import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import TextField from "./TextField";

test("basic snapshot", () => {
  const component = renderer.create(<TextField />);

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
