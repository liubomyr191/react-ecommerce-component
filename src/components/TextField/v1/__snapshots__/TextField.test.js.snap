// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<input
  className="sc-bdVaJa bPvjUk"
  maxLength={undefined}
  name="test"
  onBlur={[Function]}
  onChange={[Function]}
  onKeyPress={[Function]}
  placeholder={undefined}
  readOnly={false}
  type="text"
  value=""
/>
`;

exports[`renders textarea 1`] = `
<input
  className="sc-bdVaJa bPvjUk"
  maxLength={undefined}
  name="test"
  onBlur={[Function]}
  onChange={[Function]}
  onKeyPress={[Function]}
  placeholder={undefined}
  readOnly={false}
  type="text"
  value=""
/>
`;

exports[`renders textarea with props 1`] = `
<input
  className="CLASSNAME sc-bdVaJa bPvjUk"
  maxLength={20}
  name="test"
  onBlur={[Function]}
  onChange={[Function]}
  onKeyPress={[Function]}
  placeholder="PLACEHOLDER"
  readOnly={true}
  type="text"
  value="VALUE"
/>
`;

exports[`renders with props 1`] = `
<input
  className="CLASSNAME sc-bdVaJa bPvjUk"
  maxLength={20}
  name="test"
  onBlur={[Function]}
  onChange={[Function]}
  onKeyPress={[Function]}
  placeholder="PLACEHOLDER"
  readOnly={true}
  type="text"
  value="VALUE"
/>
`;
